### PR TITLE
`wrapreturntype`: `never`

### PR DESCRIPTION
**Pull request by mike-at-home**
This PR adds `never` as an allowed option to `wrapreturntype`, which is consistent then with `wrapeffects`.

Rationale: Codebases migrating _away from_ `wrapreturntype`: `if-multiline` and `to` `closingparen`: `balanced` will encounter issues like the following:

```
func methodName(
    arg0: Int,
    arg1: String
)
-> String
{
}
```

The new option will result in:
```
func methodName(
    arg0: Int,
    arg1: String
) -> String {
}
```

This is _probably_ what most users want by default, but I'll defer to a maintainer on that. 

@calda As we discussed earlier

### CODE REVIEW

File: Sources/FormattingHelpers.swift [hunk 1]
```diff
@@ -385,6 +394,19 @@ extension Formatter {
...
+                case .never:
+                    if startOfLine(at: endOfFunctionScope) != startOfLine(at: returnArrowIndex) {
+                        let rangeToReplace = endOfLine(at: endOfFunctionScope) ..< returnArrowIndex
+
+                        if tokens[rangeToReplace].allSatisfy(\.isSpaceOrLinebreak) {
+                            safeReplace(rangeToReplace, with: [.space(" ")])
+                        }
+                    }
+
+                    if let notWhitespace = index(of: .nonSpaceOrCommentOrLinebreak, before: openBracket) {
```
Comment (mike-at-home) [inline Sources/FormattingHelpers.swift:406]: I probably should use just `nonSpaceOrLinebreak` because this will delete comments? Though, that has some precedent I've found.

File: Sources/FormattingHelpers.swift [hunk 2]
```diff
@@ -330,16 +330,24 @@ extension Formatter {
             endOfFunctionScope: Int
         ) {
             guard token(at: startOfScope) == .startOfScope("("),
-                  let openBracket = index(of: .startOfScope("{"), after: endOfFunctionScope)
+                  var openBracket = index(of: .startOfScope("{"), after: endOfFunctionScope)
             else { return }
 
+            // adjusts openBracket to be up-to-date
```
Comment (calda) [inline Sources/FormattingHelpers.swift:336]: nice, makes sense

Comment (calda) [reply Sources/FormattingHelpers.swift:336]: although perhaps instead it makes sense to always just reinvoke `index(of: .startOfScope("{"), after: endOfFunctionScope)` every time we need the `openBracketIndex`, rather than invoking it once and keeping in sync manually

File: Sources/Options.swift [hunk 3]
```diff
@@ -112,8 +112,12 @@ public enum ExtensionACLPlacement: String, CaseIterable {
 
 /// Wrapping behavior for the return type of a function declaration
 public enum WrapReturnType: String, CaseIterable {
-    case ifMultiline = "if-multiline"
     case preserve
+    /// `-> ReturnType` is wrapped to the line after the closing paren
+    /// /// if the function spans multiple lines
```
Comment (calda) [inline Sources/Options.swift:117]:
```suggestion
    /// if the function spans multiple lines
```

File: Sources/FormattingHelpers.swift (hunk 1 above)
Comment (calda) [reply Sources/FormattingHelpers.swift:406]: Yeah, we should generally avoid deleting comments when unwrapping. I used to not think about that so didn't think to handle it in `wrapeffects`. Test cases for both `--wrapeffects never` and `--wrapreturn never` that leave the code as-is if there are comments would be good additions!

File: Sources/Options.swift [hunk 4]
```diff
@@ -112,8 +112,12 @@ public enum ExtensionACLPlacement: String, CaseIterable {
...
+    case ifMultiline = "if-multiline"
+    /// `-> ReturnType` is never wrapped, and always include on the same line as the closing paren
+    case never
```
Comment (calda) [inline Sources/Options.swift:120]: Remember to update the option description in `OptionDescriptors.swift` to mention this as an option!

File: Tests/Rules/WrapArgumentsTests.swift [hunk 5]
```diff
@@ -1802,6 +1802,61 @@ class WrapArgumentsTests: XCTestCase {
...
+            wrapReturnType: .never,
+            wrapEffects: .never
+        )
+
+        testFormatting(for: input, output, rule: .wrapArguments, options: options)
+    }
+
+    func testFormatReturnTypeOnMultilineFunctionDeclarationWithBlockComment() {
+        let input = """
+        func multilineFunction(
```
Comment (calda) [inline Tests/Rules/WrapArgumentsTests.swift:1835]: We should also handle this case:

```swift
func multilineFunction(
  foo: String,
  bar: String
)
// Comment before return
-> String
{}
```

As-is I think this only handles comments between the return type and the open `{`

File: Sources/FormattingHelpers.swift [hunk 6]
```diff
@@ -403,7 +403,10 @@ extension Formatter {
                         }
                     }
 
-                    if let notWhitespace = index(of: .nonSpaceOrCommentOrLinebreak, before: openBracket) {
+                    // search backward and replace whitespace with a single " "
+                    // if we find a line comment (// ...) do not make this change
```
Comment (calda) [inline Sources/FormattingHelpers.swift:407]: Looks like this supports both line comments and block comments based on the test cases

```suggestion
                    // if we find a comment do not make this change
```

**Comment by miguel-jimenez-0529** [comment]
Nice!

File: Sources/FormattingHelpers.swift (hunk 1 above)
Comment (mike-at-home) [reply Sources/FormattingHelpers.swift]: I've made some significant changes to this. I guess the one thing I'm unclear on is whether I should modify the existing behavior. Take a look at my latest changes and LMK if I should just ditch the `preserveComments` argument I added to the new `unwrapLine` function.

File: Tests/Rules/WrapArgumentsTests.swift (hunk 5 above)
Comment (mike-at-home) [reply Tests/Rules/WrapArgumentsTests.swift:1835]: Yeap. I was going to make both options preserve, but then I found a test case that demanded non-preservation, so... take a look at my changes and LMK your preference.

Comment (mike-at-home) [reply Tests/Rules/WrapArgumentsTests.swift:1835]: That is currently preserved, but it's a bit wonky with indentation. Is it worth being more comprehensive with weird comment and line break edge cases?

```
func multilineFunction(
    foo _: String,
    bar _: String
)
    // this is a comment
    -> String {}
```

Comment (calda) [reply Tests/Rules/WrapArgumentsTests.swift:1835]: haha yeah ok that's pretty bad, I'm fine with not supporting it in that case 👍🏻

File: Sources/Tokenizer.swift [hunk 7]
```diff
@@ -98,6 +98,8 @@ public enum TokenType {
     case number
     case error
 
+    case effectSpecifier
```
Comment (calda) [inline Sources/Tokenizer.swift:101]: I don't think we should introduce a new token type to the tokenizer, we should leave these parsed as `keyword` tokens (or `identifier` in the case of contextual keywords)

File: Sources/Tokenizer.swift [hunk 8]
```diff
@@ -397,6 +401,12 @@ public extension Token {
     var isSpaceOrComment: Bool { isSpace || isComment }
     var isSpaceOrCommentOrLinebreak: Bool { isSpaceOrComment || isLinebreak }
     var isCommentOrLinebreak: Bool { isComment || isLinebreak }
+    var isEffectSpecifier: Bool {
```
Comment (calda) [inline Sources/Tokenizer.swift:404]: I do like having a convenient helper for checking is this is an effect specifier

File: Sources/Tokenizer.swift (hunk 7 above)
Comment (calda) [inline Sources/Tokenizer.swift:101]: I'd prefer we not add this, and instead do something like `formatter.index(after: searchIndex, where: { $0.isEffectSpecifier })` at the callsite

File: Sources/FormattingHelpers.swift [hunk 9]
```diff
@@ -329,60 +329,53 @@ extension Formatter {
...
-                        // When wrapping the effect, we should also un-wrap any return type
-                        if let returnArrowIndex = funcDeclaration.returnOperatorIndex,
-                           let tokenBeforeArrowIndex = index(of: .nonSpaceOrCommentOrLinebreak, before: returnArrowIndex),
-                           startOfLine(at: tokenBeforeArrowIndex) != startOfLine(at: returnArrowIndex)
-                        {
-                            replaceTokens(in: endOfLine(at: tokenBeforeArrowIndex) ..< returnArrowIndex, with: [.space(" ")])
-                        }
+                    guard closingParenLine == startOfLine(at: effectsIndex) else { break }
+                    cursorIndex += wrapLine(before: effectsIndex)
+                    shouldUnwrapReturnArrow = true
```
Comment (calda) [inline Sources/FormattingHelpers.swift:349]: Good idea

File: Tests/Rules/WrapArgumentsTests.swift [hunk 10]
```diff
@@ -1849,6 +1849,141 @@ class WrapArgumentsTests: XCTestCase {
...
+        {
+            print("hello")
+        }
+        """
+
+        let output = """
+        func multilineFunction(
+            foo _: String,
+            bar _: String
+        ) -> String {
```
Comment (calda) [inline Tests/Rules/WrapArgumentsTests.swift:1972]: Hooray!

File: Sources/OptionDescriptor.swift [hunk 11]
```diff
@@ -586,7 +586,7 @@ struct _Descriptors {
     let wrapReturnType = OptionDescriptor(
         argumentName: "wrapreturntype",
         displayName: "Wrap Return Type",
-        help: "Wrap return type: \"if-multiline\", \"preserve\" (default)",
+        help: "Wrap return type: \"if-multiline\", \"preserve\" (default), \"never\"",
```
Comment (calda) [inline Sources/OptionDescriptor.swift:589]: This is over the 80 character limit (it's 82 chars right now), would suggest doing:

```suggestion
        help: "Wrap return type: \"if-multiline\", \"preserve\", \"never\"",
```
